## João Victor Nascimento da Silva
# Atividade Final de Machine Learning

In [1]:
import pandas as pd

Selecionei um dataset que contém dados de alunos ficticios do ensino médio estadunidense, e com base nos dados disponíveis, o objetivo é encontrar relações para conseguir prever a sua Nota final (GradeClass)

In [2]:
# Importando o Dataset
data = pd.read_csv('./Student_performance_data.csv')
data

StudentID  Age  Gender  Ethnicity  ParentalEducation  StudyTimeWeekly  \
0          1001   17       1          0                  2        19.833723   
1          1002   18       0          0                  1        15.408756   
2          1003   15       0          2                  3         4.210570   
3          1004   17       1          0                  3        10.028829   
4          1005   17       1          0                  2         4.672495   
...         ...  ...     ...        ...                ...              ...   
2387       3388   18       1          0                  3        10.680555   
2388       3389   17       0          0                  1         7.583217   
2389       3390   16       1          0                  2         6.805500   
2390       3391   16       1          1                  0        12.416653   
2391       3392   16       1          0                  2        17.819907   

      Absences  Tutoring  ParentalSupport  Extracurricular  Sports  Music  \
0            7         1                2                0       0      1   
1            0         0                1                0       0      0   
2           26         0                2                0       0      0   
3           14         0                3                1       0      0   
4           17         1                3                0       0      0   
...        ...       ...              ...              ...     ...    ...   
2387         2         0                4                1       0      0   
2388         4         1                4                0       1      0   
2389        20         0                2                0       0      0   
2390        17         0                2                0       1      1   
2391        13         0                2                0       0      0   

      Volunteering       GPA  GradeClass  
0                0  2.929196         2.0  
1                0  3.042915         1.0  
2                0  0.112602         4.0  
3                0  2.054218         3.0  
4                0  1.288061         4.0  
...            ...       ...         ...  
2387             0  3.455509         0.0  
2388             0  3.279150         4.0  
2389             1  1.142333         2.0  
2390             0  1.803297         1.0  
2391             1  2.140014         1.0  

[2392 rows x 15 columns]

### Descrição dos atributos

StudentID: Um identificador único atribuído a cada estudante (1001 a 3392).

<b>Detalhes Demográficos</b>

Idade: A idade dos estudantes varia de 15 a 18 anos.<br>
Gênero: Gênero dos estudantes, onde 0 representa Masculino e 1 representa Feminino.<br>
Etnia: A etnia dos estudantes, codificada da seguinte forma:<br>
0: Caucasiano<br>
1: Afro-Americano<br>
2: Asiático<br>
3: Outros<br>
Educação dos Pais: O nível de educação dos pais, codificado da seguinte forma:<br>
0: Nenhuma<br>
1: Ensino Médio<br>
2: Alguma Faculdade<br>
3: Bacharelado<br>
4: Superior<br>

<b>Hábitos de Estudo</b>

Tempo de Estudo Semanal: Tempo semanal de estudo em horas, variando de 0 a 20.<br>
Ausências: Número de ausências durante o ano letivo, variando de 0 a 30.<br>
Aulas de Reforço: Status de aulas de reforço, onde 0 indica Não e 1 indica Sim.<br>

<b>Envolvimento dos Pais</b>

Apoio dos Pais: O nível de apoio dos pais, codificado da seguinte forma:<br>
0: Nenhum<br>
1: Baixo<br>
2: Moderado<br>
3: Alto<br>
4: Muito Alto<br>

<b>Atividades Extracurriculares</b>

Atividades Extracurriculares: Participação em atividades extracurriculares, onde 0 indica Não e 1 indica Sim.<br>
Esportes: Participação em esportes, onde 0 indica Não e 1 indica Sim.<br>
Música: Participação em atividades musicais, onde 0 indica Não e 1 indica Sim.<br>
Voluntariado: Participação em voluntariado, onde 0 indica Não e 1 indica Sim.<br>

<b>Desempenho Acadêmico</b>

GPA: Média de notas (Grade Point Average) em uma escala de 2.0 a 4.0, influenciada pelos hábitos de estudo, envolvimento dos pais e atividades extracurriculares.<br>

<b>Variável Alvo (Target Variable): Classe de Nota</b>

Classe de Nota: Classificação das notas dos estudantes com base no GPA:<br>
0: 'A' (GPA >= 3.5)<br>
1: 'B' (3.0 <= GPA < 3.5)<br>
2: 'C' (2.5 <= GPA < 3.0)<br>
3: 'D' (2.0 <= GPA < 2.5)<br>
4: 'F' (GPA < 2.0)

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2392 entries, 0 to 2391
Data columns (total 15 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   StudentID          2392 non-null   int64  
 1   Age                2392 non-null   int64  
 2   Gender             2392 non-null   int64  
 3   Ethnicity          2392 non-null   int64  
 4   ParentalEducation  2392 non-null   int64  
 5   StudyTimeWeekly    2392 non-null   float64
 6   Absences           2392 non-null   int64  
 7   Tutoring           2392 non-null   int64  
 8   ParentalSupport    2392 non-null   int64  
 9   Extracurricular    2392 non-null   int64  
 10  Sports             2392 non-null   int64  
 11  Music              2392 non-null   int64  
 12  Volunteering       2392 non-null   int64  
 13  GPA                2392 non-null   float64
 14  GradeClass         2392 non-null   float64
dtypes: float64(3), int64(12)
memory usage: 280.4 KB


## Análise dos Dados
Correlação dos dados

In [14]:
data.drop(columns=['StudentID'], inplace=True)

In [15]:
data.corr()

Age    Gender  Ethnicity  ParentalEducation  \
Age                1.000000  0.044895  -0.028473           0.025099   
Gender             0.044895  1.000000   0.016010           0.006771   
Ethnicity         -0.028473  0.016010   1.000000           0.033595   
ParentalEducation  0.025099  0.006771   0.033595           1.000000   
StudyTimeWeekly   -0.006800  0.011469   0.007184          -0.011051   
Absences          -0.011511  0.021479  -0.025712           0.036518   
Tutoring          -0.012076 -0.031597  -0.017440          -0.017340   
ParentalSupport    0.033197  0.008065   0.020922          -0.017463   
Extracurricular   -0.025061 -0.005964  -0.008927           0.007479   
Sports            -0.046320 -0.008897  -0.004484           0.002029   
Music             -0.003492  0.007109  -0.014627           0.039439   
Volunteering       0.013074 -0.000200   0.013468           0.011960   
GPA                0.000275 -0.013360   0.027760          -0.035854   
GradeClass        -0.006250  0.022998  -0.023326           0.041031   

                   StudyTimeWeekly  Absences  Tutoring  ParentalSupport  \
Age                      -0.006800 -0.011511 -0.012076         0.033197   
Gender                    0.011469  0.021479 -0.031597         0.008065   
Ethnicity                 0.007184 -0.025712 -0.017440         0.020922   
ParentalEducation        -0.011051  0.036518 -0.017340        -0.017463   
StudyTimeWeekly           1.000000  0.009326  0.028930         0.035800   
Absences                  0.009326  1.000000 -0.015534         0.002108   
Tutoring                  0.028930 -0.015534  1.000000        -0.000824   
ParentalSupport           0.035800  0.002108 -0.000824         1.000000   
Extracurricular          -0.022860  0.000360  0.004865        -0.008381   
Sports                    0.006836  0.041454  0.006278        -0.006176   
Music                     0.007791 -0.008692 -0.011385         0.035122   
Volunteering             -0.016604 -0.018528 -0.050898        -0.006036   
GPA                       0.179275 -0.919314  0.145119         0.190774   
GradeClass               -0.134131  0.728633 -0.111695        -0.136823   

                   Extracurricular    Sports     Music  Volunteering  \
Age                      -0.025061 -0.046320 -0.003492      0.013074   
Gender                   -0.005964 -0.008897  0.007109     -0.000200   
Ethnicity                -0.008927 -0.004484 -0.014627      0.013468   
ParentalEducation         0.007479  0.002029  0.039439      0.011960   
StudyTimeWeekly          -0.022860  0.006836  0.007791     -0.016604   
Absences                  0.000360  0.041454 -0.008692     -0.018528   
Tutoring                  0.004865  0.006278 -0.011385     -0.050898   
ParentalSupport          -0.008381 -0.006176  0.035122     -0.006036   
Extracurricular           1.000000 -0.011820 -0.014191     -0.007427   
Sports                   -0.011820  1.000000 -0.020474     -0.002799   
Music                    -0.014191 -0.020474  1.000000      0.017224   
Volunteering             -0.007427 -0.002799  0.017224      1.000000   
GPA                       0.094078  0.057859  0.073318      0.003258   
GradeClass               -0.069733 -0.026654 -0.036065      0.013156   

                        GPA  GradeClass  
Age                0.000275   -0.006250  
Gender            -0.013360    0.022998  
Ethnicity          0.027760   -0.023326  
ParentalEducation -0.035854    0.041031  
StudyTimeWeekly    0.179275   -0.134131  
Absences          -0.919314    0.728633  
Tutoring           0.145119   -0.111695  
ParentalSupport    0.190774   -0.136823  
Extracurricular    0.094078   -0.069733  
Sports             0.057859   -0.026654  
Music              0.073318   -0.036065  
Volunteering       0.003258    0.013156  
GPA                1.000000   -0.782835  
GradeClass        -0.782835    1.000000

Como podemos ver, os dados com mais relação com a coluna GradeClass são:<br>
Absences/Ausencias<br>
GPA (Inversamente proporcional)

Distribuição das várias com mais relação

In [23]:
data.groupby('Absences').size()/data.shape[0]

Absences
0     0.028010
1     0.035535
2     0.028846
3     0.033027
4     0.028846
5     0.034699
6     0.036789
7     0.035953
8     0.038043
9     0.028010
10    0.025502
11    0.029682
12    0.038462
13    0.038880
14    0.035117
15    0.036371
16    0.033027
17    0.033863
18    0.033027
19    0.037625
20    0.037625
21    0.034699
22    0.029682
23    0.035953
24    0.033863
25    0.038880
26    0.033445
27    0.035117
28    0.022575
29    0.028846
dtype: float64

In [24]:
data.groupby('GPA').size()/data.shape[0]

GPA
0.000000    0.006689
0.012063    0.000418
0.019798    0.000418
0.021156    0.000418
0.026469    0.000418
              ...   
3.947093    0.000418
3.951460    0.000418
3.979421    0.000418
3.984674    0.000418
4.000000    0.002926
Length: 2371, dtype: float64

Como podemos ver elas estão bem distribuidas

## Preparação dos dados
Scaling / normalização

In [17]:
from sklearn.preprocessing import MinMaxScaler
min_max_scaler = MinMaxScaler()
data_min_max_scaled = pd.DataFrame(min_max_scaler.fit_transform(data), columns=data.columns)

In [20]:
data_min_max_scaled.head()

Age  Gender  Ethnicity  ParentalEducation  StudyTimeWeekly  Absences  \
0  0.666667     1.0   0.000000               0.50         0.992773  0.241379   
1  1.000000     0.0   0.000000               0.25         0.771270  0.000000   
2  0.000000     0.0   0.666667               0.75         0.210718  0.896552   
3  0.666667     1.0   0.000000               0.75         0.501965  0.482759   
4  0.666667     1.0   0.000000               0.50         0.233840  0.586207   

   Tutoring  ParentalSupport  Extracurricular  Sports  Music  Volunteering  \
0       1.0             0.50              0.0     0.0    1.0           0.0   
1       0.0             0.25              0.0     0.0    0.0           0.0   
2       0.0             0.50              0.0     0.0    0.0           0.0   
3       0.0             0.75              1.0     0.0    0.0           0.0   
4       1.0             0.75              0.0     0.0    0.0           0.0   

        GPA  GradeClass  
0  0.732299        0.50  
1  0.760729        0.25  
2  0.028151        1.00  
3  0.513555        0.75  
4  0.322015        1.00

## Treinamento e Validação